### PR TITLE
040 login system adam b

### PR DESCRIPTION
**Pull request by AdamB-Kainos**
finished login page frontend

### CODE REVIEW

**Review by ThomasS-kainos — APPROVED** [review]

**Review by ThomasS-kainos — APPROVED** [review]

**Review by ThomasS-kainos — APPROVED** [review]

File: .pa11yci [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
     "urls": [
-
+
```
Comment (shaunganley) [inline .pa11yci]: add the login URL

File: src/main/controllers/AuthControllers.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
+import { LoginRequest } from "../models/LoginRequest";
+import { getTokenByloggingIn} from "../services/AuthService";
+import express from "express";
+
+export const getLoginForm = async (req: express.Request, res: express.Response): Promise<void> => {
+    res.render('pages/loginPage.html', {pageName: 'Login page'});
```
Comment (shaunganley) [inline src/main/controllers/AuthControllers.ts:6]: don't need to pass the page name in here as we won't be re-using the login page for anything else

File: src/main/controllers/AuthControllers.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        res.redirect('/logoutForm')
+    } catch (e) {
+        res.locals.errormessage = e.message;
+        res.locals.pageName = "Login Page"
+        res.render('pages/loginPage.html', req.body);
+    }
+}
+
+export const getLogoutForm = async (req: express.Request, res: express.Response): Promise<void> => {
+    res.render('pages/logoutPage.html', {pageName: 'Logout page'});
```
Comment (shaunganley) [inline src/main/controllers/AuthControllers.ts:22]: don't need to pass the page name in here as we won't be re-using the logout page for anything else

File: src/main/controllers/AuthControllers.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+export const getLoginForm = async (req: express.Request, res: express.Response): Promise<void> => {
+    res.render('pages/loginPage.html', {pageName: 'Login page'});
+}
+
+export const postLoginForm  = async (req: express.Request, res: express.Response): Promise<void> => {
+    try {
+        const loginRequest: LoginRequest = req.body;
+        req.session.token = await getTokenByloggingIn(loginRequest);
+        res.redirect('/logoutForm')
```
Comment (shaunganley) [inline src/main/controllers/AuthControllers.ts]: are we able to redirect to a homepage yet?

File: src/main/models/User.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export type User = {
+    email: string,
+    salt: string,
+    hash: string,
```
Comment (shaunganley) [inline src/main/models/User.ts]: I don't think you need the salt and hash in the frontend

File: src/main/services/AuthService.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import axios, { AxiosResponse } from "axios";
+import { LoginRequest } from "../models/LoginRequest";
+
+axios.defaults.baseURL = process.env.API_URL || 'http://localhost:8080/';
+
+export const URL: string = "/api/auth/login";
+
+export const getTokenByloggingIn = async (loginRequest: LoginRequest): Promise<string> => {
+    try{
+        const response: AxiosResponse = await axios.post("http://localhost:8080/api/auth/login", loginRequest);
```
Comment (shaunganley) [inline src/main/services/AuthService.ts]: use the URL string from above, otherwise this won't work on AWS

File: src/main/models/User.ts [hunk 7]
```diff
@@ -0,0 +1,6 @@
+export type User = {
```
Comment (shaunganley) [inline src/main/models/User.ts]: also, I don't think this is used

File: src/test/unit/controllers/AuthContollerTests.ts [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+      const loginRequestObj: LoginRequest = {
+        email: "adam@random.com",
+        password: "pass123"
+      }
+
+      sinon.stub(AuthService, 'getTokenByloggingIn').resolves('12345');
+
+      const req = {
+        body: loginRequestObj,
+        session: { token: '12345' }
```
Comment (shaunganley) [inline src/test/unit/controllers/AuthContollerTests.ts]: don't set the token to that here. instead expect that the token has that value after calling the postLoginForm mtehod

File: src/test/unit/controllers/AuthContollerTests.ts [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+      sinon.stub(AuthService, 'getTokenByloggingIn').resolves('12345').rejects(new Error(errorMessage));
+
+      const invalidEmailLoginRequestObj: LoginRequest = {
+        email: "invalid@random.com",
+        password: "pass123"
+      }
+
+      const req = {
+        body: invalidEmailLoginRequestObj,
+        session: { token: '12345' }
```
Comment (shaunganley) [inline src/test/unit/controllers/AuthContollerTests.ts]: don't set the token to that here. instead expect that the token is empty after calling the postLoginForm method

File: src/test/unit/controllers/AuthContollerTests.ts [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+        locals: { errormessage: '' }
+      };
+
+      await AuthController.postLoginForm(req as express.Request, res as unknown as express.Response);
+
+      expect(res.render.calledOnce).to.be.true;
+      expect(res.locals.errormessage).to.equal(errorMessage);
+    });
+
+    it('should render loginForm view with  User is not valid: Invalid Credentials. when said error is thrown', async () => {
```
Comment (shaunganley) [inline src/test/unit/controllers/AuthContollerTests.ts]: this tests the same code as above, remove

File: src/test/unit/controllers/AuthContollerTests.ts [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+      const res = {
+        render: sinon.spy(),
+        redirect: sinon.spy()
+      };
+
+      await AuthController.postLogoutForm(req as express.Request, res as unknown as express.Response);
+
+      expect(res.redirect.calledOnce).to.be.true;
+      expect(res.redirect.calledWith('/loginForm')).to.be.true;
```
Comment (shaunganley) [inline src/test/unit/controllers/AuthContollerTests.ts]: expect token to. be empty here as well

File: src/test/unit/services/AuthServiceTests.ts [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+describe('AuthService', function () {
+
+    describe('getToken', function () {
+      it('should get Jwt token', async () => {
+        mock.onPost(URL).reply(200);
+
+        const result = await getTokenByloggingIn(loginResquest);
+
+        expect(result).toString.length > 100;
```
Comment (shaunganley) [inline src/test/unit/services/AuthServiceTests.ts]: what is this checking? you should be returning a token from the mock and expecting that value to be returned from the service call

**Review by shaunganley — CHANGES_REQUESTED** [review]

File: src/test/unit/controllers/AuthContollerTests.ts (hunk 8 above)
Comment (ThomasS-kainos) [reply src/test/unit/controllers/AuthContollerTests.ts]: Sorry but I'm lost with this. If I don't define the session token this error occurs:
     TypeError: Cannot set properties of undefined (setting 'errormessage')

It returns the stub value but when we try to assign it to req.session.token it breaks.

I imaging we are setting the wrong variable though I don't know which one we need to set

Comment (shaunganley) [reply src/test/unit/controllers/AuthContollerTests.ts]: set the value as empty string or null or undefined then expect the value to be set to '12345' by the method call

File: .pa11yci (hunk 1 above)
Comment (ThomasS-kainos) [reply .pa11yci]: Are we fine to use the AWS url or do we need to make a GitHub action to generate this file and store the url as a secret?

Comment (shaunganley) [reply .pa11yci]: use the AWS URL for running the checks on github but we might need to look into a separate set of URLs for local testing as well but that's a problem for another time :)

File: src/test/unit/controllers/AuthContollerTests.ts [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+      };
+
+      const res = {
+        render: sinon.spy(),
+        redirect: sinon.spy(),
+      };
+
+      await AuthController.postLoginForm(req as express.Request, res as unknown as express.Response);
+
+      expect(res.redirect.calledOnce).to.be.true;
```
Comment (shaunganley) [inline src/test/unit/controllers/AuthContollerTests.ts]: verify token is now '12345'

File: src/test/unit/controllers/AuthContollerTests.ts [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+      expect(res.redirect.calledWith('/')).to.be.true;
+    });
+
+    it('should render loginForm view with  User does not exist error message when said error is thrown', async () => {
+      const errorMessage: string = 'User does not exist.';
+      sinon.stub(AuthService, 'getTokenByloggingIn').resolves('12345').rejects(new Error(errorMessage));
```
Comment (shaunganley) [inline src/test/unit/controllers/AuthContollerTests.ts]: shouldn't resolve and reject

**Review by shaunganley — APPROVED** [review]

**Pull request merged**